doc/man/nsdb-delete-fsl.8
.\"@(#)nsdb-delete-fsl.8"
.\"
.\" @file doc/man/nsdb-delete-fsl.8
.\" @brief man page for nsdb-delete-fsl client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH NSDB-DELETE-FSL 8 "@publication-date@"
.SH NAME
nsdb-delete-fsl \- delete a fileset location (FSL) record from an NSDB
.SH SYNOPSIS
.B nsdb-delete-fsl
.RB [ \-?d ]
.RB [ \-D
.IR binddn ]
.RB [ \-e
.IR nce ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.I fsl-uuid
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
The bulk of FedFS junction information in a FedFS domain is stored
on one or more LDAP servers.
These servers are known as
.IR "namespace databases" ,
or NSDBs, for short.
.P
FedFS-enabled file servers and clients access the information stored
on NSDBs via standard LDAP queries.
FedFS-enabled file servers use these queries to resolve FedFS junctions.
FedFS administrators use them to manage information
about file sets contained in a FedFS domain name space.
.SH DESCRIPTION
The
.BR nsdb-delete-fsl (8)
command is part of a collection of low-level single-use programs
that is intended for testing the NSDB protocol or for use in scripts.
This command deletes a FedFS
.I fileset location
(FSL) record from an NSDB.
.P
A fileset location, or FSL, uniquely identifies the location of one
replica of a fileset.
An FSL consists of two UUIDs, the hostname and port of an NSDB,
and the hostname and port of the file server where the replica resides.
The meaning of these items is described in more detail in
.BR fedfs (7).
.P
FSLs are stored in records on an NSDB.
These records are stored as children of FSN records.
Replicas of these records can exist on more than one LDAP server.
.P
The
.BR nsdb-delete-fsl (8)
command removes an FSL record from the named NSDB.
It does not remove parent FSN records.
To remove FSN records, use the
.BR nsdb-delete-fsn (8)
command.
It does not remove a replica of a fileset.
To remove fileset replicas,
use appropriate file server administrative commands.
.P
This command has two positional parameters.
The first parameter specifies the target FSN UUID.
If a record for this FSN does not already exist, the
.BR nsdb-delete-fsn (8)
command  fails.
The second  parameter specifies the UUID of the FSL record to remove.
If a record for this FSL does not exist, the
.B nsdb-delete-fsn (8)
command fails.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR nsdb-delete-fsl (8)
version information and a usage message on
.IR stderr .
.IP "\fB-D, \-\-binddn=\fIbind-distinguished-name\fP"
Specifies a distinguished name of an entity used to bind to the LDAP server
where the NSDB resides.
If the
.B \-\-binddn
option is not specified,
the value of the FEDFS_NSDB_ADMIN environment variable is consulted.
If this variable is not set,
the NSDB connection parameter database is searched for this DN.
If none of these is specified, or
if this entity does not have permission to modify this area
of the server's DIT, the
.BR nsdb-delete-fsl (8)
command fails.
.IP "\fB-e, \-\-nce=\fINSDB-container-entry-distinguished-name\fP"
Specifies the distinguished name of the NSDB Container Entry
under which this FSL is to be created.
If the
.B \-\-nce
option is not specified,
the value of the FEDFS_NSDB_NCE environment variable is consulted.
If this variable is not set,
then the NSDB connection parameter database is searched for this DN.
If none of these is specified, the
.BR nsdb-delete-fsl (8)
command fails.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-hostname\fP"
Specifies the hostname of the NSDB where the new FSL record should reside.
If the
.B \-\-nsdbname
option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B \-\-nsdbname
option is not specified, the
.BR nsdb-delete-fsl (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the IP port of the NSDB where the new FSL record should reside.
If the
.B \-\-nsdbport
option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.SH EXIT CODES
The NSDB returns a value that reflects the success of the requested operation.
.TP
.B FEDFS_OK
The LDAP modify request succeeded.
.TP
.B FEDFS_ERR_ACCESS
The bound entity does not have permission to perform the requested operation.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The
.BR nsdb-delete-fsl (8)
command was unable to find a route to the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The
.BR nsdb-delete-fsl (8)
command determined that the specified NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The
.BR nsdb-delete-fsl (8)
command was unable to establish a connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The
.BR nsdb-delete-fsl (8)
command was unable to authenticate
and establish a secure connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
A non-specific LDAP error occurred on the connection between the
.BR nsdb-delete-fsl (8)
command and specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
An LDAP error occurred on the connection between the
.BR nsdb-delete-fsl (8)
command and specified NSDB.
The specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_NONCE
The
.BR nsdb-delete-fsl (8)
command was unable to locate the NCE on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSN
The
.BR nsdb-delete-fsl (8)
command was unable to locate the specified FSN on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSL
The
.BR nsdb-delete-fsl (8)
command was unable to locate the specified FSL for the specified FSN
on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The
.BR nsdb-delete-fsl (8)
command received a malformed response from the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the specified NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The local NSDB connection parameter database
does not have any connection parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The
.BR nsdb-delete-fsl (8)
command received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
The
.BR nsdb-delete-fsl (8)
command received an LDAP referral that it was unable to follow.
A specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The
.BR nsdb-delete-fsl (8)
command received an LDAP referral that it chose not to follow,
either because the local implementation does not support
following LDAP referrals or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The
.BR nsdb-delete-fsl (8)
command received an LDAP referral that it chose not to follow
because the local NSDB connection parameter database had no
connection parameters for the NSDB targeted by the LDAP referral.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that you have an FSN that looks like:
.RS
.sp
	FSN UUID: 8e246ddc-7b46-11e0-8252-000c297fd679
.br
	NSDB: nsdb.example.net:389
.sp
.RE
Further suppose the NSDB
.I nsdb.example.net:389
has an NSDB Container Entry whose distinguished name is
.IR o=fedfs .
The replica of this fileset that exists at
.I fileserver.example.net:/path
has just been removed, and it's FSL UUID is
.IR 323c5068-7c11-11e0-8d38-000c297fd679 .
To delete the corresponding FSL record, you might use:
.RS
.sp
$ nsdb-delete-fsl -D cn=Manager -e o=fedfs \\
.br
	-l nsdb.example.net \\
.br
	8e246ddc-7b46-11e0-8252-000c297fd679
.br
	323c5068-7c11-11e0-8d38-000c297fd679
.br
Enter NSDB password:
.br
Successfully deleted FSL record
  fedfsFslUuid=323c5068-7c11-11e0-8d38-000c297fd679,
  fedfsFsnUuid=8e246ddc-7b46-11e0-8252-000c297fd679,o=fedfs
.sp
.RE
The FSL record for the specified replica is removed,
leaving possibly other FSL records for this fileset,
and leaving the parent FSN record intact.
.SH SECURITY
Permission to modify the LDAP's DIT is required to delete an FSL record.
The
.BR nsdb-delete-fsl (8)
command must bind as an entity permitted to modify the DIT
to perform this operation.
The
.BR nsdb-delete-fsl (8)
command asks for a password on
.IR stdin .
Standard password blanking techniques are used
to obscure the password on the user's terminal.
.P
The target LDAP server must be registered in the local NSDB connection
parameter database.
The connection security mode listed
in the NSDB connection parameter database
for the target LDAP server is used during this operation.
See
.BR nsdbparams (8)
for details on how to register an NSDB
in the local NSDB connection parameter database.
.SH "SEE ALSO"
.BR fedfs (7),
.BR nsdb-delete-fsn (8),
.BR nsdb-list (8),
.BR nsdbparams (8)
.sp
RFC 5716 for FedFS requirements and overview
.sp
RFC 4510 for an introduction to LDAP
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/nsdbc/nsdb-delete-fsl.c
/**
 * @file src/nsdbc/nsdb-delete-fsl.c
 * @brief Delete a FedFS FSL entry from a target NSDB server
 *
 * @todo
 *	Add some logic to find the full DN given only the FSL uuid.
 *	For now, both the FSN and FSL uuids are required.
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>
#include <netdb.h>
#include <langinfo.h>

#include <uuid/uuid.h>

#include "fedfs.h"
#include "nsdb.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Short form command line options
 */
static const char nsdb_delete_fsl_opts[] = "?dD:e:l:r:";

/**
 * Long form command line options
 */
static const struct option nsdb_delete_fsl_longopts[] = {
	{ "binddn", 1, NULL, 'D', },
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "nce", 1, NULL, 'e', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nsdbport", 1, NULL, 'r', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 */
static void
nsdb_delete_fsl_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [ -d ] [ -D binddn ] "
			"[ -l nsdbname ] [ -r nsdbport ] [ -e nce ] "
			"fsl-uuid\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-D, --binddn         Bind DN\n");
	fprintf(stderr, "\t-e, --nce            DN of NSDB container entry\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *binddn, *nsdbname;
	char *nce, *fsl_uuid;
	unsigned short nsdbport;
	unsigned int ldap_err;
	FedFsStatus retval;
	nsdb_t host;
	int arg;

	(void)umask(S_IRWXO);

	/* Ensure UTF-8 strings can be handled transparently */
	if (setlocale(LC_CTYPE, "") == NULL ||
	    strcmp(nl_langinfo(CODESET), "UTF-8") != 0) {
		fprintf(stderr, "Failed to set locale and langinfo\n");
		exit((int)FEDFS_ERR_INVAL);
	}

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb_env(&nsdbname, &nsdbport, &binddn, &nce);

	while ((arg = getopt_long(argc, argv, nsdb_delete_fsl_opts,
			nsdb_delete_fsl_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'D':
			binddn = optarg;
			break;
		case 'e':
			nce = optarg;
			break;
		case 'l':
			nsdbname = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdbport)) {
				fprintf(stderr, "Bad port number: %s\n",
					optarg);
				nsdb_delete_fsl_usage(progname);
			}
			break;
		default:
			fprintf(stderr, "Invalid command line "
				"argument: %c\n", (char)arg);
		case '?':
			nsdb_delete_fsl_usage(progname);
		}
	}
	if (argc == optind + 1) {
		uuid_t uu;
		fsl_uuid = argv[optind];
		if (uuid_parse(fsl_uuid, uu) == -1) {
			fprintf(stderr, "Invalid FSL UUID was specified\n");
			nsdb_delete_fsl_usage(progname);
		}
	} else if (argc > optind + 1) {
		fprintf(stderr, "Unrecognized positional parameters\n");
		nsdb_delete_fsl_usage(progname);
	} else {
		fprintf(stderr, "No FSL UUID was specified\n");
		nsdb_delete_fsl_usage(progname);
	}
	if (nsdbname == NULL) {
		fprintf(stderr, "No NSDB hostname was specified\n");
		nsdb_delete_fsl_usage(progname);
	}

	retval = nsdb_lookup_nsdb(nsdbname, nsdbport, &host, NULL);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_PARAMS:
		fprintf(stderr, "No connection parameters for NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out;
	default:
		fprintf(stderr, "Failed to look up NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out;
	}
	retval = FEDFS_ERR_INVAL;
	if (binddn == NULL)
		binddn = (char *)nsdb_default_binddn(host);
	if (binddn == NULL) {
		fprintf(stderr, "No NDSB bind DN was specified\n");
		goto out_free;
	}
	if (nce == NULL)
		nce = (char *)nsdb_default_nce(host);
	if (nce == NULL) {
		fprintf(stderr, "No NCE was specified\n");
		goto out_free;
	}

	retval = nsdb_open_nsdb(host, binddn, NULL, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_CONN:
		fprintf(stderr, "Failed to connect to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_AUTH:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u: %s\n",
			nsdbname, nsdbport, ldap_err2string(ldap_err));
		goto out_free;
	default:
		fprintf(stderr, "Failed to bind to NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out_free;
	}

	retval = nsdb_delete_fsl_s(host, nce, fsl_uuid, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		printf("Successfully deleted FSL record for %s under %s\n",
				fsl_uuid, nce);
		break;
	case FEDFS_ERR_NSDB_NONCE:
		if (nce == NULL)
			fprintf(stderr, "NSDB %s:%u has no NCE\n",
				nsdbname, nsdbport);
		else
			fprintf(stderr, "NCE %s does not exist\n", nce);
		break;
	case FEDFS_ERR_NSDB_NOFSL:
		fprintf(stderr, "NSDB %s:%u has no such FSL %s\n",
			nsdbname, nsdbport, fsl_uuid);
		break;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "Failed to delete FSL %s: %s\n",
			fsl_uuid, ldap_err2string(ldap_err));
		break;
	default:
		fprintf(stderr, "Failed to delete FSL %s: %s\n",
			fsl_uuid, nsdb_display_fedfsstatus(retval));
	}

	nsdb_close_nsdb(host);

out_free:
	nsdb_free_nsdb(host);

out:
	exit((int)retval);
}
